core/opendaq/synchronization/include/opendaq/sync_component_impl.h

```diff
@@ -0,0 +1,308 @@
+
+    PropertyObjectPtr interfacePtr = syncInterface;
+
+    //TBD: Check if interface inherits from SyncInterfaceBase
+    StringPtr className = interfacePtr.getClassName();
+    if (className == "SyncInterfaceBase")
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, "Allowed adding property objects that inherit from 'SyncInterfaceBase', but not 'SyncInterfaceBase' itself.");
+    }
+
+    auto typeManager = this->context.getTypeManager();
+    if (typeManager == nullptr)
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_ARGUMENT_NULL, "TypeManager is not assigned.");
+    }
+
+    TypePtr type;
+    ErrCode errCode = typeManager->getType(className, &type);
+    if (OPENDAQ_FAILED(errCode) || type == nullptr)
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' not found.", className));
+    }
+
+    if (auto objectClass = type.asPtrOrNull<IPropertyObjectClass>(true); objectClass.assigned())
+    {
+        auto parentName = objectClass.getParentName();
+        if (!parentName.assigned() || parentName != "SyncInterfaceBase")
+        {
+            return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' does not inherit from 'SyncInterfaceBase'.", className));
+        }
+    }
+    else
+    {
+        return this->makeErrorInfo(OPENDAQ_ERR_INVALID_ARGUMENT, fmt::format("Interface '{}' is not IPropertyObjectClass", className));
+    }
+
+    BaseObjectPtr interfacesValue;
+    StringPtr str = "Interfaces";
+    ErrCode err = this->getPropertyValue(str, &interfacesValue);
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    const auto interfacesPtr = interfacesValue.asPtr<IPropertyObject>(true);
+    err = interfacesPtr->addProperty(ObjectProperty(className, syncInterface));
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    return daqTry([&]() {
+        ListPtr<IString> interfaceNames = getTypedProperty<IList>("InterfaceNames");
+        interfaceNames.pushBack(className);
+        return Super::setPropertyValue(String("InterfaceNames"), interfaceNames);
+    });
+}
+
+
+template <typename MainInterface, typename ... Interfaces>
+ErrCode GenericSyncComponentImpl<MainInterface, Interfaces...>::removeInterface(IString* interfaceName)
+{
+    OPENDAQ_PARAM_NOT_NULL(interfaceName);
+
+    BaseObjectPtr interfacesValue;
+    StringPtr str = "Interfaces";
+    ErrCode err = this->getPropertyValue(str, &interfacesValue);
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    const auto InterfacesPtr = interfacesValue.asPtr<IPropertyObject>(true);
+    err = InterfacesPtr->removeProperty(interfaceName);
+    if (OPENDAQ_FAILED(err))
+        return err;
+
+    return daqTry([&]() {
+        Int selectedSource = 0;
+        getSelectedSource(&selectedSource);
+
+        ListPtr<IString> interfaceNames = getTypedProperty<IList>("InterfaceNames");
+        for (SizeT i = 0; i < interfaceNames.getCount(); i++)
```

Comment:
What happens when the list is empty?

Reply:
nothing will happened as we are initilizing "InterfaceNames" in constructor with empty list. which in native passes while deserilizing. In opcua its populated in tmsSyncComponent constructor as well

Reply:
Is something to consider in the future, but this implies that anyone using selection properties needs to double check that the index is not out of bounds.